<a href="https://colab.research.google.com/github/Meaiie/Data_viz_Midterm/blob/main/Midterm_exam.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exam 1



## Import ข้อมูล

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import os # os ใช้จัดการ path ของ file / ใช้เชื่อม path กับ file

path_to = '/content/drive/MyDrive/Data_viz/test-TidyData_2024.xlsx'

In [ ]:
Tidy_data = pd.read_excel(path_to)
Tidy_data

month  station  rep.    Chlo_a    Chlo_b  total_Caro  Temp.    pH  \
0    July_2018        1     1  0.094645  0.011954    0.026780   30.2  5.54   
1    July_2018        2     1  0.038433  0.059019    0.021387   29.1  6.24   
2    July_2018        3     1  0.043884  0.089415    0.023223   29.7  5.52   
3    July_2018        4     1  0.039836  0.064552    0.016537   29.2  5.92   
4    July_2018        5     1  0.044712  0.062424    0.022990   29.5  5.79   
..         ...      ...   ...       ...       ...         ...    ...   ...   
193  July_2019        2     3  0.030521  0.015358    0.018067   31.2  7.74   
194  July_2019        3     3  0.024081  0.020217    0.015981   29.7  7.50   
195  July_2019        4     3  0.015525  0.009612    0.009373   31.5  7.51   
196  July_2019        5     3  0.015180  0.011811    0.009697   28.5  7.31   
197  July_2019        6     3  0.008556  0.010605    0.005704   31.5  7.71   

        EC  Salinity    TDS    DO   Nitrate  Ammonium  Phosphate  \
0    622.7     330.3  545.2  4.37  0.172902  0.052817   0.012748   
1    580.7     307.0  488.2  6.55  0.154197  0.144843   0.006086   
2    547.1     289.4  462.2  5.67  0.153097  0.154321   0.014158   
3    473.6     249.2  397.9  0.69  0.157026  0.387408   0.056951   
4    685.5     366.0  581.2  4.46  0.204024  0.043387   0.007239   
..     ...       ...    ...   ...       ...       ...        ...   
193  580.5     309.2  517.0  6.31  0.132034  0.030959  -0.000064   
194  513.0     272.6  436.0  3.00  0.230116  0.059103  -0.000064   
195  550.9     289.8  498.4  4.02  0.123546  0.136428  -0.000064   
196  540.6     287.4  443.4  1.53  0.110028  0.056057  -0.000064   
197  544.7     288.2  493.8  6.32  0.165042  0.105817  -0.000064   

     No.Algal_cell  @dropdown  
0     1.635833e+05        NaN  
1     1.166667e+06        NaN  
2     2.970000e+06        NaN  
3     5.066667e+05        NaN  
4     1.265000e+06        NaN  
..             ...        ...  
193   2.665833e+05        NaN  
194   6.428333e+05        NaN  
195   2.945000e+05        NaN  
196   3.470833e+05        NaN  
197   3.692500e+05        NaN  

[198 rows x 17 columns]

## 1. ความแตกต่างของ ค่าเฉลี่ย ของข้อมูล Chlo_a ทุกๆ station ระหว่างปี 2018 และ ปี 2019

In [ ]:
Tidy_data.isnull().sum()

month              0
station            0
rep.               0
Chlo_a             0
Chlo_b             0
total_Caro         0
Temp.              0
pH                 0
EC                 0
Salinity           0
TDS                0
DO                 0
Nitrate            0
Ammonium           0
Phosphate          0
No.Algal_cell      0
@dropdown        198
dtype: int64

In [ ]:
# สร้างคอลัมน์ year เพิ่ม
Tidy_data['year'] = Tidy_data['month'].str[-4:]
Tidy_data


month  station  rep.    Chlo_a    Chlo_b  total_Caro  Temp.    pH  \
0    July_2018        1     1  0.094645  0.011954    0.026780   30.2  5.54   
1    July_2018        2     1  0.038433  0.059019    0.021387   29.1  6.24   
2    July_2018        3     1  0.043884  0.089415    0.023223   29.7  5.52   
3    July_2018        4     1  0.039836  0.064552    0.016537   29.2  5.92   
4    July_2018        5     1  0.044712  0.062424    0.022990   29.5  5.79   
..         ...      ...   ...       ...       ...         ...    ...   ...   
193  July_2019        2     3  0.030521  0.015358    0.018067   31.2  7.74   
194  July_2019        3     3  0.024081  0.020217    0.015981   29.7  7.50   
195  July_2019        4     3  0.015525  0.009612    0.009373   31.5  7.51   
196  July_2019        5     3  0.015180  0.011811    0.009697   28.5  7.31   
197  July_2019        6     3  0.008556  0.010605    0.005704   31.5  7.71   

        EC  Salinity    TDS    DO   Nitrate  Ammonium  Phosphate  \
0    622.7     330.3  545.2  4.37  0.172902  0.052817   0.012748   
1    580.7     307.0  488.2  6.55  0.154197  0.144843   0.006086   
2    547.1     289.4  462.2  5.67  0.153097  0.154321   0.014158   
3    473.6     249.2  397.9  0.69  0.157026  0.387408   0.056951   
4    685.5     366.0  581.2  4.46  0.204024  0.043387   0.007239   
..     ...       ...    ...   ...       ...       ...        ...   
193  580.5     309.2  517.0  6.31  0.132034  0.030959  -0.000064   
194  513.0     272.6  436.0  3.00  0.230116  0.059103  -0.000064   
195  550.9     289.8  498.4  4.02  0.123546  0.136428  -0.000064   
196  540.6     287.4  443.4  1.53  0.110028  0.056057  -0.000064   
197  544.7     288.2  493.8  6.32  0.165042  0.105817  -0.000064   

     No.Algal_cell  @dropdown  year  
0     1.635833e+05        NaN  2018  
1     1.166667e+06        NaN  2018  
2     2.970000e+06        NaN  2018  
3     5.066667e+05        NaN  2018  
4     1.265000e+06        NaN  2018  
..             ...        ...   ...  
193   2.665833e+05        NaN  2019  
194   6.428333e+05        NaN  2019  
195   2.945000e+05        NaN  2019  
196   3.470833e+05        NaN  2019  
197   3.692500e+05        NaN  2019  

[198 rows x 18 columns]

In [ ]:
# Group the data by 'Station' and 'Year', then calculate the mean of 'Chlo_a' for each group
grouped_data = Tidy_data.groupby(['station', 'year'])['Chlo_a'].mean().reset_index()

# Pivot the table to have years as columns and stations as rows
pivot_table = grouped_data.pivot(index='station', columns='year', values='Chlo_a')

# Calculate the difference between the mean Chlo_a values for 2018 and 2019
pivot_table['Difference'] = pivot_table['2019'] - pivot_table['2018']

for idx, value in pivot_table['Difference'].items():
  if value > 0:
     pivot_table.loc[idx, 'Difference'] = value
  else:
     pivot_table.loc[idx, 'Difference'] = -value

pivot_table

year         2018      2019  Difference
station                                
1        0.052529  0.018268    0.034261
2        0.048112  0.027911    0.020201
3        0.053285  0.023057    0.030228
4        0.046794  0.021620    0.025173
5        0.043432  0.020100    0.023331
6        0.043809  0.015594    0.028215

## 2. สร้างตารางค่าเฉลี่ย Chlo_b ทุกๆ station ของแต่ละเดือน

In [ ]:
# prompt: สร้างตารางค่าเฉลี่ย Chlo_b ทุกๆ station ของแต่ละเดือน

# Group the data by 'Station' and 'Month', then calculate the mean of 'Chlo_b' for each group
grouped_data_chlo_b = Tidy_data.groupby(['station', 'month'])['Chlo_b'].mean().reset_index()

# Pivot the table to have months as columns and stations as rows
pivot_table_chlo_b = grouped_data_chlo_b.pivot(index='station', columns='month', values='Chlo_b')

pivot_table_chlo_b


month    Apr_2019  Aug_2018  Dec_2018  Jan_2019  July_2018  July_2019  \
station                                                                 
1        0.051381  0.075665  0.070061  0.022451   0.046688   0.014365   
2        0.019378  0.065403  0.064375  0.014968   0.062731   0.013691   
3        0.030775  0.065179  0.080867  0.017722   0.073561   0.017439   
4        0.020170  0.061289  0.061951  0.012142   0.061395   0.009257   
5        0.027015  0.060107  0.070085  0.009789   0.064623   0.011054   
6        0.025655  0.063027  0.065663  0.013856   0.056193   0.010073   

month    June_2019  May_2019  Nov_2018  Oct_2018  Sep_2018  
station                                                     
1         0.020548  0.015949  0.075937  0.131764  0.063890  
2         0.013194  0.028765  0.064552  0.122767  0.074578  
3         0.013301  0.015890  0.097963  0.106404  0.076292  
4         0.013821  0.015559  0.070605  0.090834  0.064670  
5         0.013880  0.016623  0.063866  0.082381  0.060733  
6         0.008583  0.023882  0.073360  0.079449  0.061750

# EXAM 2

## import ข้อมูล

In [1]:
import pandas as pd
import os

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
path_crime_scene = '/content/drive/MyDrive/Data_viz/EXAM_DATA/crime_scene_report.csv'

In [5]:
crime_scene_report = pd.read_csv(path_crime_scene)
crime_scene_report

date     type                                        description  \
0     20180115  robbery  A Man Dressed as Spider-Man Is on a Robbery Spree   
1     20180115   murder                  Life? Dont talk to me about life.   
2     20180115   murder  Mama, I killed a man, put a gun against his he...   
3     20180215   murder                         REDACTED REDACTED REDACTED   
4     20180215   murder  Someone killed the guard! He took an arrow to ...   
...        ...      ...                                                ...   
1223  20180430  bribery                                                 \n   
1224  20180430    fraud                  ‘Why not?’ said the March Hare.\n   
1225  20180430  assault                                                 \n   
1226  20180501  assault  be NO mistake about it: it was neither more no...   
1227  20180115   murder  Security footage shows that there were 2 witne...   

              city  
0              NYC  
1           Albany  
2             Reno  
3         SQL City  
4         SQL City  
...            ...  
1223  Garden Grove  
1224         Houma  
1225       Fontana  
1226       Trenton  
1227      SQL City  

[1228 rows x 4 columns]

In [20]:
path_drivers_license = '/content/drive/MyDrive/Data_viz/EXAM_DATA/drivers_license.csv'
path_facebook_event_checkin = '/content/drive/MyDrive/Data_viz/EXAM_DATA/facebook_event_checkin.csv'
path_income = '/content/drive/MyDrive/Data_viz/EXAM_DATA/income.csv'
path_interview = '/content/drive/MyDrive/Data_viz/EXAM_DATA/interview.csv'
path_person = '/content/drive/MyDrive/Data_viz/EXAM_DATA/person.csv'
path_get_fit_now_check_in = '/content/drive/MyDrive/Data_viz/EXAM_DATA/get_fit_now_check_in.csv'
path_get_fit_now_member = '/content/drive/MyDrive/Data_viz/EXAM_DATA/get_fit_now_member.csv'

## ให้สืบหาฆาตรกร ในคดี ฆาตรกรรม (murder) ที่เกิดขึ้นในวันที่ 15 มกราคม 2018 ที่เมือง SQL City โดยเริ่มจากการตรวจสอบ ข้อมูลคดี (crime_scene_report.csv)

In [18]:
crime_scene = crime_scene_report[(crime_scene_report['date'] == 20180115)&
                                 (crime_scene_report['type'] == 'murder') &
                                 (crime_scene_report['city'] == 'SQL City')]
crime_scene

date    type                                        description  \
1227  20180115  murder  Security footage shows that there were 2 witne...   

          city  
1227  SQL City

In [28]:
pd.set_option('display.max_colwidth', None)
crime_scene['description']

1227    Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".
Name: description, dtype: object

จากภาพจากกล้องวงจรปิดพบว่ามีพยาน 2 คน พยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบน "Northwestern Dr" พยานคนที่สองชื่อแอนนาเบล อาศัยอยู่ที่ไหนสักแห่งบนถนนแฟรงคลิน

In [30]:
person = pd.read_csv(path_person)
person

id                 name  license_id  address_number  \
0      10000   Christoper Peteuil      993845             624   
1      10007  Kourtney Calderwood      861794            2791   
2      10010            Muoi Cary      385336             741   
3      10016          Era Moselle      431897            1987   
4      10025         Trena Hornby      550890             276   
...      ...                  ...         ...             ...   
10006  99936          Luba Benser      274427             680   
10007  99941      Roxana Mckimley      975942            1613   
10008  99965      Cherie Zeimantz      287627            3661   
10009  99982          Allen Cruse      251350            3126   
10010  99990         Vance Hunten      830407            3056   

      address_street_name        ssn  
0            Bankhall Ave  747714076  
1           Gustavus Blvd  477972044  
2         Northwestern Dr  828638512  
3           Wood Glade St  614621061  
4           Daws Hill Way  223877684  
...                   ...        ...  
10006        Carnage Blvd  685095054  
10007             Gate St  512136801  
10008       The Water Ave  362877324  
10009           N Jean Dr  348734531  
10010       Lancefield St  896677562  

[10011 rows x 6 columns]

เนื่องจากพยานคนแรกอาศัยอยู่ที่บ้านหลังสุดท้ายบน "Northwestern Dr" ไม่แน่ใจว่า บ้านหลังสุดท้ายในที่นี้จะระบุได้อย่างไร เนื่องจากหมายเลขบ้านไม่ได้สื่อถึงการมาก่อนหลัง และไม่สามารถบอกได้ว่าอยู่ตำแหน่งใดจึงเลือกเป็นบ้านหลังสุดท้ายของ record มา เพราะคิดว่าเป็นบ้านที่ทำการขึ้นทะเบียนมาทีหลังสุดแทน แต่ว่าเมื่อเช็คไปแล้วบ้านเลขที่นั้นไม่ได้มีในหมายเลขการสัมภาษณ์ จึงเลือกเป็นคนที่มี address_number สูงที่สุดมาแทน

In [62]:
witness_1 = person[person['address_street_name'] == 'Northwestern Dr']
witness_1 = witness_1[witness_1['address_number'] == witness_1['address_number'].max()]
witness_1

id            name  license_id  address_number address_street_name  \
499  14887  Morty Schapiro      118009            4919     Northwestern Dr   

           ssn  
499  111564949

เนื่องจากพยานคนที่สองชื่อแอนนาเบล อาศัยอยู่ที่ไหนสักแห่งบนถนนแฟรงคลิน

In [47]:
witness_2 = person[(person['address_street_name'] == 'Franklin Ave') &
                   (person['name'].str.contains('Annabel'))]

witness_2

id            name  license_id  address_number address_street_name  \
665  16371  Annabel Miller      490173             103        Franklin Ave   

           ssn  
665  318771143

- การสันนิษฐานครั้งที่ 1
จะพบว่าพยานคนที่ 1 คือ Ivy Kazarian (License_id	994470) และพยานคนที่ 2 คือ Annabel Miller (License_id 490173)

- การสันนิษฐานครั้งที่ 2
จะพบว่าพยานคนที่ 1 คือ Morty Schapiro (id	14887) และพยานคนที่ 2 คือ Annabel Miller (id 16371)

In [29]:
interview = pd.read_csv(path_interview)
interview

person_id  \
0         28508   
1         63713   
2         86208   
3         35267   
4         33856   
...         ...   
4986      37357   
4987      10206   
4988      14887   
4989      16371   
4990      67318   

                                                                                                                                                                                                                                              transcript  
0                                                                                                                                                                                                                    ‘I deny it!’ said the March Hare.\n  
1                                                                                                                                                                                                                                                     \n  
2                                                                                                                                                                                                          way, and the whole party swam to the shore.\n  
3                                                                                                                                                                                lessons in here? Why, there’s hardly room for YOU, and no room at all\n  
4                                                                                                                                                                                                                                                     \n  
...                                                                                                                                                                                                                                                  ...  
4986                                                                                                                                                                                Alice did not wish to offend the Dormouse again, so she began very\n  
4987                                                                                                                                                                          time,’ she said, ‘than waste it in asking riddles that have no answers.’\n  
4988                     I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".  
4989                                                                                                                               I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.  
4990  I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.\n  

[4991 rows x 2 columns]

In [65]:
witness_interview = interview[interview['person_id'].isin([14887,16371])]
witness_interview

person_id  \
4988      14887   
4989      16371   

                                                                                                                                                                                                                           transcript  
4988  I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".  
4989                                                                                                            I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.

 - จากการสัมภาษณ์ Morty Schapiro
  ฉันได้ยินเสียงปืนแล้วก็เห็นชายคนหนึ่งวิ่งออกไป เขามีกระเป๋า "Get Fit Now Gym" หมายเลขสมาชิกบนกระเป๋าขึ้นต้นด้วย "48Z" เฉพาะสมาชิกระดับโกลด์เท่านั้นที่มีถุงเหล่านั้น ชายคนนั้นขึ้นรถป้ายทะเบียนที่ประกอบด้วย "H42W"
 - จากการสัมภาษณ์ Annabel Miller
  ฉันเห็นการฆาตกรรมเกิดขึ้น และฉันก็จำฆาตกรได้จากยิมของฉันตอนที่ฉันกำลังออกกำลังกายเมื่อสัปดาห์ที่แล้วในวันที่ 9 มกราคม

In [50]:
driver_license = pd.read_csv(path_drivers_license)
driver_license

id  age  height eye_color hair_color  gender plate_number  \
0      100280   72      57     brown        red    male       P24L4U   
1      100460   63      72     brown      brown  female       XF02T6   
2      101029   62      74     green      green  female       VKY5KR   
3      101198   43      54     amber      brown  female       Y5NZ08   
4      101255   18      79      blue       grey  female       5162Z1   
...       ...  ...     ...       ...        ...     ...          ...   
10002  999923   19      77     amber      black  female       5L0ZI4   
10003  999940   71      61     green      green    male       1B8QN8   
10004  999981   67      69     brown       blue  female       1684K3   
10005  999986   49      58     green       grey    male       F8F64H   
10006  999993   18      63     black      black  female       6UZO2O   

         car_make    car_model  
0           Acura          MDX  
1        Cadillac          SRX  
2           Scion           xB  
3          Nissan        Rogue  
4           Lexus           GS  
...           ...          ...  
10002         GMC  Sierra 3500  
10003  Mitsubishi      Eclipse  
10004  Land Rover          LR2  
10005       Lexus           LS  
10006    Cadillac      DeVille  

[10007 rows x 9 columns]

In [67]:
get_fit_member = pd.read_csv(path_get_fit_now_member)
get_fit_member

id  person_id                name  membership_start_date  \
0    NL318      65076     Everette Koepke               20170926   
1    AOE21      39426        Noe Locascio               20171005   
2    2PN28      63823  Jeromy Heitschmidt               20180215   
3    0YJ24      80651      Waneta Wellard               20171206   
4    3A08L      32858        Mei Bianchin               20170401   
..     ...        ...                 ...                    ...   
179  2V137      41693      Wendell Dulany               20171219   
180  4KB72      79110          Emile Hege               20170522   
181  48Z7A      28819        Joe Germuska               20160305   
182  48Z55      67318       Jeremy Bowers               20160101   
183  90081      16371      Annabel Miller               20160208   

    membership_status  
0                gold  
1             regular  
2              silver  
3                gold  
4              silver  
..                ...  
179            silver  
180           regular  
181              gold  
182              gold  
183              gold  

[184 rows x 5 columns]

In [71]:
# ตรวจสอบสมาชิกที่มี id ขึ้นต้นด้วย 48Z และเป็นสมาชิกระดับ Gold
suspect_gym = get_fit_member[(get_fit_member['membership_status'] == 'gold') & (get_fit_member['id'].str.startswith('48Z'))]
suspect_gym


id  person_id           name  membership_start_date membership_status
181  48Z7A      28819   Joe Germuska               20160305              gold
182  48Z55      67318  Jeremy Bowers               20160101              gold

พบว่ามีสมาชิกระดับ gold 2 คนที่มี id สมาชิกขึ้นต้นด้วย 48Z

จากการสัมภาษณ์ Annabel Miller ฉันเห็นการฆาตกรรมเกิดขึ้น และฉันก็จำฆาตกรได้จากยิมของฉันตอนที่ฉันกำลังออกกำลังกายเมื่อสัปดาห์ที่แล้วในวันที่ 9 มกราคม

In [70]:
get_fit_checkin = pd.read_csv(path_get_fit_now_check_in)
get_fit_checkin

membership_id  check_in_date  check_in_time  check_out_time
0            NL318       20180212            329             365
1            NL318       20170811            469             920
2            NL318       20180429            506             554
3            NL318       20180128            124             759
4            NL318       20171027            418            1019
...            ...            ...            ...             ...
2698         4KB72       20170422           1016            1114
2699         4KB72       20170630            408             885
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700
2702         90081       20180109           1600            1700

[2703 rows x 4 columns]

In [74]:
gym_20180109 = get_fit_checkin[(get_fit_checkin['check_in_date'] == 20180109)&
                               (get_fit_checkin['membership_id'].isin(suspect_gym['id']))
                               ]
gym_20180109

membership_id  check_in_date  check_in_time  check_out_time
2700         48Z7A       20180109           1600            1730
2701         48Z55       20180109           1530            1700

พบว่าทั้งคู่ได้เข้าเช็คอินวันที่ 9 มกราคม
ผู้ต้องสงสัยคนที่ 1 คือ Joe Germuska (Person id 28819)
ผู้ต้องสงสัยคนที่ 2 คือ Jeremy Bowers (Person id 67318)

In [76]:
suspect_info = person[person['id'].isin([28819,67318])]
suspect_info

id           name  license_id  address_number    address_street_name  \
2037  28819   Joe Germuska      173289             111                Fisk Rd   
6327  67318  Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

            ssn  
2037  138909730  
6327  871539279

In [66]:
driver_license[driver_license['plate_number'].str.contains('H42W')]

id  age  height eye_color hair_color  gender plate_number  \
915   183779   21      65      blue     blonde  female       H42W0X   
3529  423327   30      70     brown      brown    male       0H42W2   
6240  664760   21      71     black      black    male       4H42WR   

       car_make car_model  
915      Toyota     Prius  
3529  Chevrolet  Spark LS  
6240     Nissan    Altima

จะพบว่ามีรถอยู่ 3 คันที่มีเลขทะเบียนประกอบด้วย 'H42W' ได้แก่ id 183779 , 423327 , 664760 ซึ่งมีเลขทะเบียนที่ตรงอยู่ คือ 423327 นั่นคือรถของ Jeremy Bowers
  ดังนั้นคนร้ายคือ Jeremy Bowers

In [77]:
suspect_interview = interview[interview['person_id'] == 67318]
suspect_interview

person_id  \
4990      67318   

                                                                                                                                                                                                                                              transcript  
4990  I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.\n

Jeremy Bowers ให้การว่า ฉันถูกจ้างโดยผู้หญิงที่มีเงินมากมาย ฉันไม่รู้ชื่อของเธอ แต่ฉันรู้ว่าเธอสูงประมาณ 5'5" (65") หรือ 5'7" (67") เธอมีผมสีแดงและเธอขับรถ Tesla Model S ฉันรู้ว่าเธอเข้าร่วม SQL Symphony Concert 3 ครั้งในเดือนธันวาคม 2017

In [78]:
facebook_event_checkin = pd.read_csv(path_facebook_event_checkin)
facebook_event_checkin

person_id  event_id  \
0          28508      5880   
1          63713      3865   
2          63713      3999   
3          63713      6436   
4          82998      4470   
...          ...       ...   
20006      99716      1143   
20007      99716      1143   
20008      99716      1143   
20009      67318      4719   
20010      67318      1143   

                                                                event_name  \
0                          Nudists are people who wear one-button suits.\n   
1      but that's because it's the best book on anything for the layman.\n   
2                                 If Murphy's Law can go wrong, it will.\n   
3         Old programmers never die.  They just branch to a new address.\n   
4                                     Help a swallow land at Capistrano.\n   
...                                                                    ...   
20006                                                 SQL Symphony Concert   
20007                                                 SQL Symphony Concert   
20008                                                 SQL Symphony Concert   
20009                                               The Funky Grooves Tour   
20010                                                 SQL Symphony Concert   

           date  
0      20170913  
1      20171009  
2      20170502  
3      20170926  
4      20171022  
...         ...  
20006  20171206  
20007  20171212  
20008  20171229  
20009  20180115  
20010  20171206  

[20011 rows x 4 columns]

In [79]:
# prompt: หาคนที่เข้าร่วม SQL Symphony Concert 3 ครั้งในเดือนธันวาคม 2017 จาก facebook_event_checkin

# หาคนที่เข้าร่วม SQL Symphony Concert 3 ครั้งในเดือนธันวาคม 2017
sql_symphony_checkin = facebook_event_checkin[facebook_event_checkin['event_name'] == 'SQL Symphony Concert']
sql_symphony_checkin_dec = sql_symphony_checkin[sql_symphony_checkin['date'].between(20171201,20171231)]
checkin_count = sql_symphony_checkin_dec.groupby('person_id')['event_name'].count().reset_index(name='count')
suspect_employer = checkin_count[checkin_count['count'] == 3]
suspect_employer


person_id  count
4       24556      3
15      99716      3

In [88]:
suspect_checkin_details = facebook_event_checkin.merge(suspect_employer, left_on='person_id', right_on='person_id',
                                                      how='right')

suspect_checkin_details


person_id  event_id            event_name      date  count
0      24556      1143  SQL Symphony Concert  20171207      3
1      24556      1143  SQL Symphony Concert  20171221      3
2      24556      1143  SQL Symphony Concert  20171224      3
3      99716      1143  SQL Symphony Concert  20171206      3
4      99716      1143  SQL Symphony Concert  20171212      3
5      99716      1143  SQL Symphony Concert  20171229      3

In [94]:
suspect_2 = person[person['id'].isin([24556,99716])]
suspect_2

id              name  license_id  address_number address_street_name  \
1587  24556       Bryan Pardo      101191             703          Machine Ln   
9985  99716  Miranda Priestly      202298            1883          Golden Ave   

            ssn  
1587  816663882  
9985  987756388

มี 2 คนที่ไป SQL Symphony Concert มา 3 ครั้ง ภายในเดือนธันวาคม 2017 คือ Bryan Pardo และ Miranda Priestly

In [96]:
suspect_license = driver_license[driver_license['id'].isin([101191,202298])]
suspect_license

id  age  height eye_color hair_color  gender plate_number car_make  \
1105  202298   68      66     green        red  female       500123    Tesla   

     car_model  
1105   Model S

In [98]:
income = pd.read_csv(path_income)
income

ssn  annual_income
0     100009868          52200
1     100169584          64500
2     100300433          74400
3     100355733          35900
4     100366269          73000
...         ...            ...
7509  999679296          54400
7510  999762859          77000
7511  999824984          82000
7512  999910617          82600
7513  999942603          11500

[7514 rows x 2 columns]

In [99]:
suspect_info2 = income[income['ssn'] == 987756388]
suspect_info2

ssn  annual_income
7422  987756388         310000

พบข้อมูล license ของ Miranda Priestly ซึ่งเธอขับ Tesla Model S มีผมสีแดง สูงราว ๆ 65'ถึง 67' และมีรายได้ต่อปีที่สูงถึง 310000 $ ตรงตามคำให้การของผู้ก่อเหตุ จึงสรุปได้ว่าเธอเป็นผู้ว่าจ้าง